### ISSUE TITLE
TurboSnap not working

### ISSUE THREAD

**Issue by duarten** [chromaui/chromatic-cli#446]
TurboSnap doesn't appear to be working for my project. This always happens:

```
$ chromatic --ci --exit-once-uploaded --only-changed

Chromatic CLI v6.0.5
https://www.chromatic.com/docs/cli

Authenticating with Chromatic
    → Connecting to https://index.chromatic.com
Authenticated with Chromatic
    → Using project token '******hf36'
Retrieving git information
⚠ Ignoring --only-changed
Could not retrieve changed files since baseline commit(s).
This typically happens after rebasing.
```

This was on a fresh PR, up-to-date with the main branch.

**Comment by tmeasday** [comment]
Hey @duarten -- perhaps you could try running with the `--debug` flag? What is the commit of the previous build? Are you using github "merge" commits to run CI against?

**Comment by duarten** [comment]
Hi @tmeasday! I'll follow-up as soon as I've run chromatic with `--debug`, but regarding your other questions:

The commit of the previous build was the tip of the main branch, built on a push event.

The PR build was done in a `pull_request` event, but I'm checking out `github.event.pull_request.head.sha` instead of the merge commit.

**Comment by tmeasday** [comment]
Hey @duarten so the message "Could not retrieve changed files since baseline commit(s)." means that the previous build we are comparing to is attached to a commit that is no longer in the git history. The debug info should make it easier to tell what that build/commit is.

**Comment by duarten** [comment]
I haven't been able to reproduce the original problem. I wonder if the issue was that the build for the parent commit (at the head of the main branch) hadn't finished by the time CI called chromatic for the PR based on that parent? In any case, it would be nice it the cli accepted a list of changed files, as most CI systems already calculate those in one step or another.

In the meanwhile I've noticed this message:

```
Failed to retrieve dependent story files {"statsPath":"/tmp/chromatic--2181-cTRMO2KLssg0/preview-stats.json","changedFiles":[".github/workflows/pr-checks.yml"]}
```

It seems that in this case, all snapshots were created?

Finally, it seems named imports aren't taken into account when calculating dependencies. If I modify a component C1 in package P, then stories that import C2 from P are also detected as changed. Is this correct?

**Comment by tmeasday** [comment]
> I wonder if the issue was that the build for the parent commit (at the head of the main branch) hadn't finished by the time CI called chromatic for the PR based on that parent? 

Hmm, I don't think that would lead to this effect (looking like the commit is missing from the repo), but the opposite (not finding the "right" commit).

> Failed to retrieve dependent story files

This happens if there's an error doing the tracing process. Perhaps you could try following the debugging steps here: https://www.chromatic.com/docs/turbosnap#why-are-no-changes-being-detected

> Finally, it seems named imports aren't taken into account when calculating dependencies. If I modify a component C1 in package P, then stories that import C2 from P are also detected as changed. Is this correct?

That would happen if you have both `C1` and `C2` being exported from `P` from a single `index.js` file. That's just the information webpack gives us unfortunately.

**Comment by duarten** [comment]
> > I wonder if the issue was that the build for the parent commit (at the head of the main branch) hadn't finished by the time CI called chromatic for the PR based on that parent?
> 
> Hmm, I don't think that would lead to this effect (looking like the commit is missing from the repo), but the opposite (not finding the "right" commit).

I haven't been able to reproduce this yet :(

> 
> > Failed to retrieve dependent story files
> 
> This happens if there's an error doing the tracing process. Perhaps you could try following the debugging steps here: https://www.chromatic.com/docs/turbosnap#why-are-no-changes-being-detected

It seems the issue is due to the changed files not being in the webpack build. Would it be possible to just ignore those files? I think they're triggering a full-retest.

> 
> > Finally, it seems named imports aren't taken into account when calculating dependencies. If I modify a component C1 in package P, then stories that import C2 from P are also detected as changed. Is this correct?
> 
> That would happen if you have both `C1` and `C2` being exported from `P` from a single `index.js` file. That's just the information webpack gives us unfortunately.

Got it! Thanks.

**Comment by tmeasday** [comment]
> It seems the issue is due to the changed files not being in the webpack build. Would it be possible to just ignore those files? I think they're triggering a full-retest.

Hmm, no, if webpack doesn't know about the files, it would lead to an "empty" build. This is actually a common case in a monorepo, where changes can occur in non-UI subprojects.

**Comment by duarten** [comment]
No it's not possible to ignore them, or no a full-retest isn't happening? :)

**Comment by tmeasday** [comment]
No, I wouldn't expect a retest to happen, sorry that was unclear.

**Comment by duarten** [comment]
Closing this as after all this time the error that prompted me to open this problem never reproduced again :)

**Comment by robhung** [comment]
Hello - I'm experiencing a similar issue and have some output with the `--debug` flag.

```
  | 2023-02-02 00:55:24 UTC | Authenticating with Chromatic
  | 2023-02-02 00:55:24 UTC | → Connecting to https://index.chromatic.com
  | 2023-02-02 00:55:24 UTC | Authenticated with Chromatic
  | 2023-02-02 00:55:24 UTC | → Using project token '***********'
  | 2023-02-02 00:55:24 UTC | Retrieving git information
  | 2023-02-02 00:55:24 UTC | git info: {"commit":"257cd60702e845025361f8e9af45417d1a3bca62","committedAt":1675298750000,"committerEmail":"robert@dovetailapp.com","committerName":"Robert Hung","branch":"robhung/make-it-savematic","slug":"heydovetail/platform","isTravisPrBuild":false,"fromCI":true,"ciService":"buildkite"}
  | 2023-02-02 00:55:24 UTC | ⚠ Using outdated package
  | 2023-02-02 00:55:24 UTC | You are using an outdated version of the Chromatic CLI (v5.9.2).
  | 2023-02-02 00:55:24 UTC | A new major version is available (v6.15.0).
  | 2023-02-02 00:55:24 UTC | Run yarn upgrade chromatic --latest to upgrade.
  | 2023-02-02 00:55:24 UTC | App firstBuild: '{"committedAt":1628816610000}', lastBuild: '{"commit":"cce5ca129e320d6b9288ff2447fac4494c22b8b3","committedAt":1675297462000}', pullRequest: 'null'
  | 2023-02-02 00:55:24 UTC | Adding last branch build commit cce5ca129e320d6b9288ff2447fac4494c22b8b3 to commits with builds
  | 2023-02-02 00:55:24 UTC | step: checking 20 up to 1628816610
  | 2023-02-02 00:55:24 UTC | step: commitsWithBuilds: cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | step: commitsWithoutBuilds:
  | 2023-02-02 00:55:24 UTC | running git rev-list HEAD       --since 1628816610       -n 20 --not cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | command output: 257cd60702e845025361f8e9af45417d1a3bca62,5eaa87c1d64f6d9d5177566ae5ee4bc404a03fa0
  | 2023-02-02 00:55:24 UTC | step: candidateCommits: 257cd60702e845025361f8e9af45417d1a3bca62,5eaa87c1d64f6d9d5177566ae5ee4bc404a03fa0
  | 2023-02-02 00:55:24 UTC | step: newCommitsWithBuilds:
  | 2023-02-02 00:55:24 UTC | step: checking 40 up to 1628816610
  | 2023-02-02 00:55:24 UTC | step: commitsWithBuilds: cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | step: commitsWithoutBuilds: 257cd60702e845025361f8e9af45417d1a3bca62,5eaa87c1d64f6d9d5177566ae5ee4bc404a03fa0
  | 2023-02-02 00:55:24 UTC | running git rev-list HEAD       --since 1628816610       -n 42 --not cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | command output: 257cd60702e845025361f8e9af45417d1a3bca62,5eaa87c1d64f6d9d5177566ae5ee4bc404a03fa0
  | 2023-02-02 00:55:24 UTC | step: candidateCommits:
  | 2023-02-02 00:55:24 UTC | step: no candidateCommits; we are done
  | 2023-02-02 00:55:24 UTC | Final commitsWithBuilds: cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | running git rev-list cce5ca129e320d6b9288ff2447fac4494c22b8b3 --not "cce5ca129e320d6b9288ff2447fac4494c22b8b3^@"
  | 2023-02-02 00:55:24 UTC | command output: cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | Found parentCommits: cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | Found baselineCommits: cce5ca129e320d6b9288ff2447fac4494c22b8b3
  | 2023-02-02 00:55:24 UTC | ⚠ Ignoring --only-changed
  | 2023-02-02 00:55:34 UTC | Could not retrieve changed files since baseline commit(s).
  | 2023-02-02 00:55:34 UTC | This typically happens after rebasing.
  | 2023-02-02 00:55:34 UTC | {"timedOut":true,"signal":"SIGTERM","originalMessage":"Timed out","shortMessage":"Command timed out after 10000 milliseconds: git diff --name-only cce5ca129e320d6b9288ff2447fac4494c22b8b3\nTimed out","command":"git diff --name-only cce5ca129e320d6b9288ff2447fac4494c22b8b3","escapedCommand":"git diff --name-only cce5ca129e320d6b9288ff2447fac4494c22b8b3","signalDescription":"Termination","stdout":"","stderr":"","all":"","failed":true,"isCanceled":false,"killed":false}
  | 2023-02-02 00:55:34 UTC | Retrieved git information
  | 2023-02-02 00:55:34 UTC | → Commit '257cd60' on branch 'robhung/make-it-savematic'; found 1 parent commit
  | 2023-02-02 00:55:34 UTC | Collecting Storybook metadata
  | 2023-02-02 00:55:34 UTC | Collected Storybook metadata
  | 2023-02-02 00:55:34 UTC | → Storybook v^6.5.9 for React; supported addons found: A11y, Actions, Knobs, Viewport
  | 2023-02-02 00:55:34 UTC | Publish your built Storybook
  | 2023-02-02 00:55:34 UTC | → Validating Storybook files
  | 2023-02-02 00:55:34 UTC | Publishing your built Storybook
  | 2023-02-02 00:55:34 UTC | → Retrieving target location
  | 2023-02-02 00:55:34 UTC | → Starting publish
```

And here are my commits:

<img width="783" alt="Screenshot 2023-02-02 at 12 08 31" src="https://user-images.githubusercontent.com/11624465/216206148-6dbcbff6-1c29-482d-b094-c40c6e79d004.png">

`5eaa87c1d64f6d9d5177566ae5ee4bc404a03fa0` didn't have a build because I cancelled it early to add the debug flag, but I expected `cce5ca129e320d6b9288ff2447fac4494c22b8b3` to be the baseline commit.

From the logs, it looks like it's correctly identified the baseline commit, but just unable to retrieve the files. Any ideas?

I'm building storybook in a separate command using:

```
build-storybook --webpack-stats-json -c .storybook --output-dir storybook-build
```

We've got a monorepo using both yarn workspaces and nx - if that makes a difference?

**Comment by tmeasday** [comment]
Hi @robhung. 

- You are on quite an old version of Chromatic there, I think it's worth upgrading before going any further.
- The error is coming from running `git diff --name-only cce5ca129e320d6b9288ff2447fac4494c22b8b3` -- it looks like it is timing out, and we are assuming that means the commit doesn't exist. What happens when you run that command?

**Comment by robhung** [comment]
@tmeasday thanks for your quick assistance!

Turns out, our CI is configured with the bare minimum for `git` to run. I did a test by running `git status` before Chromatic, and it took 7 seconds to retrieve the status for the commit that was checked out.

Once I did that, TurboSnap worked with no issues.

I assume when running `git diff`, it will need to pull the files for both commits, which probably requires 14 seconds, but the timeout set for all your [git commands is 10 seconds](https://github.com/chromaui/chromatic-cli/blob/eef1d33dab80a09258877709d7902818431bc6ea/bin-src/git/git.ts#L18).

Is there a way we can pass an override to the timeout by any chance?
